pczt: Define the structure and semantics of the PCZT format

str4d opened the pull request.

str4d commented on pczt/src/common.rs:
```diff
@@ -0,0 +1,51 @@
+
+/// Global fields that are relevant to the transaction as a whole.
+#[derive(Clone)]
+pub(crate) struct Global {
```
Keystone's proposal included a `network` field (indicating mainnet vs testnet vs ?) at the global level, possibly because in my 2019 PCZT draft I suggested it. However, I note that PSBTs (BIP 174) do not include such a field (AFAICT). Is such a field necessary, and/or useful?

arya2 replied:
I don't see why it would be necessary or useful here, but it looks like BIP 174 does include an implicit indication of the network in the serialized extended public keys.

soralit replied:
We meet a similar problem on Unisat's Fractal Bitcoin recently. PSBT doesn't have a network field, so we have to set it on the global map field which means that becomes a custom protocol between Unisat and us. 
This can be useful if the chain have multiple sub/child chains like Bitcoin vs Litecoin or Dogecoin. In theory Litecoin can also use PSBT,  but without network field, we cannot know some accurate info like coin precision.
Anyway, that's depends on Zcash's needs.

pacu replied:
If a Derivation Path is provided wouldn't that account for having `network` field?

str4d replied:
Hmm, this is a good point. And as @arya2 notes above and @pacu notes below, we will already be semi-explicitly encoding the network into the PCZT at some later stage. So I think it makes sense for the Creator to encode the network up-front, and then every later instance of data that includes network information can be checked against it for correctness.

Next question then is what format this network field should have. A SLIP 44 `coin_type` (as used in BIP 32 and ZIP 32 derivation paths) would make sense for mainnet applications, but SLIP 44 only defines a single testnet code for all coins. Would that be sufficient for Zcash (and potentially other) testnet uses? (I think so, because if you're running against testnet then you already likely know what network you're dealing with, and the important part is that we have domain separation between mainnet and non-mainnet, as well as between Zcash and non-Zcash mainnets.)

str4d replied:
I decided to add a SLIP 44 `coin_type` field.

daira commented on pczt/src/common.rs:
```diff
@@ -0,0 +1,44 @@
+    // These are required fields that are part of the final transaction, and are filled in
+    // by the Creator when initializing the PCZT.
+    //
+    pub(crate) tx_version: u32,
+    pub(crate) version_group_id: u32,
```
Should this pair of fields be optional independently of the others?

arya2 replied:
Would it be better to infer the version group id from the tx version instead of having it as an explicit field?

> Should this pair of fields be optional independently of the others?

Would it make sense to adding a builder struct where the `tx_version` field is optional instead of making it optional here?

str4d replied:
> Would it be better to infer the version group id from the tx version instead of having it as an explicit field?

That's more fragile across consensus forks, but I can see an argument that the tuple `(tx_version, consensus_branch_id)` might be sufficient to fully constrain the value of `version_group_id` such that every role can reliably deduce it. The main argument against this would be that it would force PCZT roles to be aware of more details about transaction formats, where they might otherwise just be able to focus on the PCZT.

daira replied:
IMHO the arguments for encoding the version group ID explicitly in the transaction format apply equally to PCZTs.

daira commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,283 @@
+    ///
+    /// This is initialized by the Creator, and updated by the Constructor as spends or
+    /// outputs are added to the PCZT. It enables per-spend and per-output values to be
+    /// redacted from the PCZT after they are no longer necessary.
```
Maybe calculate this at the same time as bsk.

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,279 @@
+    /// This is initialized by the Creator, and updated by the Constructor as spends or
+    /// outputs are added to the PCZT. It enables per-spend and per-output values to be
+    /// redacted from the PCZT after they are no longer necessary.
+    pub(crate) value_balance: u64,
```
Same comment as Orchard.

arya2 commented on pczt/src/common.rs:
```diff
@@ -0,0 +1,44 @@
+    pub(crate) consensus_branch_id: u32,
+    /// TODO: In PSBT this is `fallback_lock_time`; decide whether this should have the
+    /// same semantics.
+    pub(crate) lock_time: u32,
+    pub(crate) expiry_height: u32,
```
It could be nice to type these as [`LockTime`](https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/transaction/lock_time.rs#L25) and `BlockHeight`.

> Should these be independently optional?

If this needs to be able to produce v1 or v2 transactions, then at least `expiry_height` should probably be optional.

str4d replied:
@arya2 I'm intentionally _not_ adding types to the core structure (at least at this stage), to keep its dependencies as minimal / non-existent as possible (because there are many roles that do not need to care about every type in the PCZT, e.g. an Orchard signer doesn't need to be able to parse Sapling-specific types). The Transaction Extractor role necessarily performs all conversions and checks, so that acts as a backstop.

Once we've figured out and agreed on the structure, then it may make sense to add certain newtypes that `pczt` can rely on (depending on who the downstream consumers of the crate will be).

daira replied:
@arya2 I don't think it does need to be able to produce v1 or v2 transactions, or v3 for that matter. I only know of use cases for using PCZTs to produce new transactions, which would necessarily have to be at least v4.

str4d replied:
- `lock_time: u32` is now `fallback_lock_time: Option<u32>` and has the same semantics as in BIP 370.
- `expiry_height` remains a `u32` and is set by the Creator. (It could in theory be altered by an Updater prior to the IO Finalizer or Signers running, but we require it to always be set.)

arya2 commented on pczt/src/common.rs:
```diff
@@ -0,0 +1,44 @@
+    ///
+    /// Non-optional because this commits to the set of consensus rules that will apply to
+    /// the transaction; differences therein can affect every role.
+    pub(crate) consensus_branch_id: u32,
```
It could be nice to have a `network_upgrade` field instead and then infer the consensus branch id from the network upgrade, or to type this as a [`ConsensusBranchId`](https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/parameters/network_upgrade.rs#L148).

str4d replied:
See above. This might make sense for the Rust-specific usage (again, depending on how carefully we can manage dependencies to ensure the types are still useful to `pczt` crate consumers), but for the general PCZT format we need to think of this as being a `u32`, so I'm keeping it this way for now.

arya2 commented on pczt/src/roles/creator/mod.rs:
```diff
@@ -0,0 +1,60 @@
+            consensus_branch_id,
+            expiry_height,
+            // Spends and outputs enabled.
+            orchard_flags: 0b0000_0011,
```
It could be nice to use bitflags here (see [`orchard::Flags`](https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/orchard/shielded_data.rs#L227) in `zebra-chain`).

arya2 commented on pczt/src/lib.rs:
```diff
@@ -1 +1,150 @@
+
+/// A partially-created Zcash transaction.
+#[derive(Clone)]
+pub struct Pczt {
```
It's very verbose, and the documentation is helpful for explaining what this is to someone that's not already familiar, but could this be renamed to `TransactionParts` or something that's more immediately understandable without context?

str4d replied:
- `UnsignedTransaction` is incorrect; the PCZT format can contain data that on its own cannot fully produce an unsigned transaction.
- `PartialTransaction` might be more reasonable (as a shortening of `PartiallyCreatedZcashTransaction`), but it doesn't clarify in what way the transaction is partial.
- `PartiallyCreatedTransaction` would be another option (just expanding `Pczt` and dropping the superfluous Zcash for which we already have the context from the `pczt` crate name).

In any case, this choice is immaterial to the discussion at hand, which is the PCZT format semantics.

arya2 commented on pczt/src/roles/creator/mod.rs:
```diff
@@ -0,0 +1,64 @@
+
+use crate::{Pczt, V5_TX_VERSION, V5_VERSION_GROUP_ID};
+
+pub struct Creator {
```
Could we link to the struct it's building in the struct docs?

Nitpick: Could we rename this to `PcztBuilder` so it's more immediately obvious that it's following the builder pattern and that it's for the other struct?

str4d replied:
This is not a PCZT builder; this is a struct implementing the Creator role, specifically as intended to follow the (applicable) roles introduced in BIP 174 and BIP 370 for PSBTs.

The "PCZT builder" is in effect spread across the Creator and Constructor roles, and arguably the Updater role as well.

pacu commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,428 @@
+    /// - After`zkproof` / `spend_auth_sig` has been set, this can be redacted.
+    pub(crate) alpha: Option<[u8; 32]>,
+
+    // TODO derivation path
```
wouldn't this account for `network` information (see comments on top of the PR)

conradoplg commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,428 @@
+
+    // TODO derivation path
+
+    // TODO FROST
```
For FROST I think we need:

- Maybe `SigningPackage` (commitments and message, in this case, the sighash which will need to be recalculated and checked by the receiver). We could do the other way around though; send the PCZT inside the message field of the `SigningPackage` and replace it with the sighash before proceeding. Or just send the two together (this is what I planned to do in the demo repo, in [a struct](https://github.com/ZcashFoundation/frost-zcash-demo/blob/3338161fc10c13a025c8e45e84903a6f3acd16da/server/src/types.rs#L107) that contains both the SigningPackage and the PCZT in the `aux_msg` field)
- Possibly `randomizer_seed` which is used to rederive `alpha` and which will be checked by the receiver, per current https://github.com/zcash/zips/pull/895. But similarly this could be done outside the PCZT.

I'm not sure we need to use PCZTs for the round 2 reply from Participants to Coordinator; it would be just a gigantic wrapper for a signature share where everything but the signature share would be thrown away anyway.

str4d replied:
> I'm not sure we need to use PCZTs for the round 2 reply from Participants to Coordinator; it would be just a gigantic wrapper for a signature share where everything but the signature share would be thrown away anyway.

You may not need to in some cases, but I can imagine workflows where the round is asynchronous and the coordinator needs to learn which signing session the signature share(s) belong to. PCZT gives you that for free: the Combiner merges the PCZT results from each participant into a single PCZT, and then the Spend Finalizer runs the FROST aggregation step.

conradoplg reviewed (commented):
Looks good, added some comments about FROST usage

conradoplg commented:
For ZSA issuance I think we need [the fields required for computing the `issuance_digest`](https://zips.z.cash/zip-0227#t-5-issuance-digest)?

str4d commented:
[Force-pushed](https://github.com/zcash/librustzcash/compare/6ca070b657e8d4b6f3530b638145f5b18b88c1e1..eb4275fbe3325fc2cafcd12ae812dfca236f5a2d) to fill in some more parts:
- Transparent bundle fields for signing (adapting from BIPs 174 and 370)
- Optional ZIP 32 paths (that Updaters can set to help Signers determine which keys to use)
- Decided on semantics for merging data in proprietary maps (equal keys are required to have equal values, same as every other field and map).

str4d commented:
I'm currently inclined to leave out both FROST support and ZSA issuance support from the first version of PCZTs, as we don't have experience with either of them yet, and trying to get them right at the same time as the base PCZT format is likely to end in sorrow.

As @conradoplg notes, we can in some cases have FROST internally pass around a PCZT, but that won't necessarily help for all possible PCZT use cases, so we'll likely want a v2 that adds those fields. That would also be the right time to adjust the core fields if we find weaknesses in the v1 PCZT format.

conradoplg commented:
I'm happy with leaving FROST out of v1, we can make it work for now by sending additional data along with the PCZT.

str4d commented:
[Force-pushed](https://github.com/zcash/librustzcash/compare/eb4275fbe3325fc2cafcd12ae812dfca236f5a2d..37fc6ce6db577167313f13faeac9c6b04561d327) with more changes:
- Anchors are now required and set by the Creator.
- Ciphertexts are now `Vec<u8>` to accommodate upcoming plaintext changes.
- `Pczt` struct no longer contains a `Version`; instead that will be explicitly part of the encoding.
- Extracted some conversion logic from the Prover role to use it for consistency checks in the Signer role.

str4d commented:
[Force-pushed](https://github.com/zcash/librustzcash/compare/37fc6ce6db577167313f13faeac9c6b04561d327..d807c7ed7f886c38e195263ae0589b3d202076f7):
- Added an IO Finalizer role, and started on a Constructor role.
- Noted some bugs in the format that need to be fixed.
- Started writing an end-to-end test.

Use of the Constructor role interacts heavily with the per-protocol crates, so I'm now working on the production impl of those, both to fix the format bugs and get us to a working client-side implementation.

str4d commented:
[Force-pushed](https://github.com/zcash/librustzcash/compare/d807c7ed7f886c38e195263ae0589b3d202076f7..243294106fec31aa5a398c33b949c6419a1d71ee):
- Migrated to the protocol-specific PCZT implementations in the `orchard` and `sapling-crypto` crates.
- Added a similar transparent PCZT impl in `zcash_primitives`.
- Fixed a bunch of bugs in the format uncovered by writing end-to-end tests.
- Wrote several end-to-end tests and got them passing.

I have a couple more changes to make to the `Global` struct (specifically it is missing the global flags), and then I'll add the `postcard` derives.

daira commented on pczt/src/lib.rs:
```diff
@@ -1 +1,135 @@
+//! - Decide on an anchor
+//!   - All spends should use the same anchor for indistinguishability
+//!   - In a future transaction version, all spends will be required to do so
+//! - Collect each spend
+//!   - Proofs can and should be created at this time
+//! - Create proofs for each spend and output
+//!   - Data necessary for proofs can be stripped out of the format
```
Are these actually separate steps?

str4d replied:
I've removed this content; it was only informational and not kept up-to-date.

daira commented on pczt/src/lib.rs:
```diff
@@ -1 +1,135 @@
+//!     flag is True. Outputs may only be added if the Outputs Modifiable flag is True.
+//!   - A single entity is likely to be both a Creator and Constructor.
+//! - IO Finalizer (anyone can execute)
+//!   - Sets the appropriate bits in PSBT_GLOBAL_TX_MODIFIABLE to 0. (TODO fix up)
```
Is this TODO able to be resolved now?

str4d replied:
No, I need to do this today.

str4d replied:
This is now resolved.

nuttycom commented on pczt/src/sapling.rs:
```diff
@@ -359,221 +349,164 @@ impl Bundle {
+            .spends()
+            .iter()
+            .map(|spend| {
+                let rcm = spend.rseed().and_then(|rseed| {
+                    if let sapling::Rseed::BeforeZip212(rcm) = rseed {
+                        Some(rcm.to_bytes())
+                    } else {
+                        None
+                    }
+                });
+                let rseed = spend.rseed().and_then(|rseed| {
+                    if let sapling::Rseed::AfterZip212(rseed) = rseed {
+                        Some(rseed)
+                    } else {
+                        None
+                    }
+                });
```
This might be clearer as a single expression that returns a tuple.

str4d commented:
[Rebased on `main`](https://github.com/zcash/librustzcash/compare/243294106fec31aa5a398c33b949c6419a1d71ee..1dd51082314ad8beb043ca050c78f3b2e7518c5a) to fix merge conflict.

str4d commented:
[Force-pushed](https://github.com/zcash/librustzcash/compare/1dd51082314ad8beb043ca050c78f3b2e7518c5a..754317db51c1319a3f097847f184ff5dd1accb2a) to address a comment and fix various lints.

str4d commented:
[Force-pushed](https://github.com/zcash/librustzcash/compare/754317db51c1319a3f097847f184ff5dd1accb2a..01155ff935620d612d93ef464f4c9b65cc54e730) to fix a feature flag combination and update dependency vets.

str4d commented:
[Force-pushed](https://github.com/zcash/librustzcash/compare/01155ff935620d612d93ef464f4c9b65cc54e730..d8e8865e5f3096aacd2cfd4fb4ea3cae9be7e775) with the remaining fixes to `Global`:
- Fixed the `lock_time` field to be `fallback_lock_time`, and implemented the BIP 370 behaviour.
- Added a `coin_type` field (addressing feedback).
- Added a `tx_modifiable` bitfield.

At this point, I believe the only thing missing is the encoding, which I'll do next (after dinner).

nuttycom commented on pczt/src/common.rs:
```diff
@@ -0,0 +1,194 @@
+}
+
+impl Global {
+    pub(crate) fn merge(mut self, other: Self) -> Option<Self> {
```
nonblocking nit: is it a problem that we lose information about what failed to merge by using `Option` here instead of `Result`?

str4d replied:
If it becomes a problem, we can change it later; this is all `pub(crate)` so isn't part of the public API.

nuttycom commented on pczt/src/transparent.rs:
```diff
@@ -0,0 +1,388 @@
+    /// - The pubkeys should appear in `script_pubkey` or `redeem_script`.
+    /// - Each entry is set by an Updater.
+    /// - Individual entries may be required by a Signer.
+    pub(crate) bip32_derivation: BTreeMap<[u8; 33], Zip32Derivation>,
```
It's a little bit odd to me that this uses ZIP 32 derivation; for something like signing transactions that spend outputs to legacy `zcashd` wallet UTXOs where the spending key was drawn from randomness rather than ZIP 32 derivation, would the process be to just look up from the pubkey and ignore the value in the map?

str4d replied:
In that case, the map would be empty. There is no need to put anything in the map; it's purely to help the Signer figure out where the key material is, but if they already know then there's no need.

daira replied:
I suggested a comment to that effect.

nuttycom commented on pczt/src/transparent.rs:
```diff
@@ -0,0 +1,388 @@
+    /// - The pubkeys should appear in `script_pubkey` or `redeem_script`.
+    /// - Each entry is set by an Updater.
+    /// - Individual entries may be required by a Signer.
+    pub(crate) bip32_derivation: BTreeMap<[u8; 33], Zip32Derivation>,
```
I feel like I'm missing something here, why is this part of the `Output`?

daira replied:
I would also like to know this!

daira replied:
I think it's essentially the same reason as for inputs, i.e. there is no other way to identify the derivation paths for keys in `script_pubkey` or `redeem_script`.

nuttycom commented on zcash_primitives/src/transaction/components/transparent/pczt.rs:
```diff
@@ -0,0 +1,214 @@
+    /// - The pubkeys should appear in `script_pubkey` or `redeem_script`.
+    /// - Each entry is set by an Updater.
+    /// - Individual entries may be required by a Signer.
+    pub(crate) bip32_derivation: BTreeMap<[u8; 33], Bip32Derivation>,
```
Same question here, what is this used for on the output side?

str4d replied:
I suspect the intention (from BIP 174) was to enable Signers to verify that they can indeed produce signatures for change outputs (and therefore verify that the change is going to an address they can subsequently spend from).

nuttycom reviewed (commented):
Flushing comments, there are a couple of things about the transparent bits that I don't understand.

str4d commented:
[Force-pushed](https://github.com/zcash/librustzcash/compare/d8e8865e5f3096aacd2cfd4fb4ea3cae9be7e775..79baf8ea24da2cbe8add60b73611c73c96d317d6) to add PCZT encoding.

nuttycom commented on pczt/Cargo.toml:
```diff
@@ -11,3 +11,67 @@ license.workspace = true
+zcp-builder = ["dep:zcash_primitives", "dep:zcash_protocol"]
+io-finalizer = ["orchard", "sapling"]
+prover = ["dep:rand_core", "sapling?/temporary-zcashd"]
+signer = ["dep:blake2b_simd", "dep:rand_core", "orchard", "sapling", "transparent"]
+spend-finalizer = ["transparent"]
+tx-extractor = ["dep:rand_core", "orchard", "sapling", "transparent"]
```
These feature flags are not sufficiently granular, but I don't know what the right way is to do it - for example, you can't use `zcash_client_backend` with the `orchard` flag disabled but also use `pczt/signer` because the latter pulls in the `orchard` dependencies.

nuttycom replied:
For now I think that I'm going to make `zcash_client_backend/pczt` imply `orchard` and `transparent-inputs` and then we can worry about other combinations later.

str4d replied:
Yep. The granularity here is limited due both to `zcash_primitives` not having flags yet, and due to the transparent protocol living in that crate. It causes a bit of a mess when you're trying to either use transparent or any role that requires multiple protocols.

nuttycom commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,524 @@
+    /// - This is set by the Constructor.
+    /// - This is required by the Prover.
+    #[serde_as(as = "Option<[_; 43]>")]
+    pub(crate) recipient: Option<[u8; 43]>,
```
There's an annoyance here with the `zcash_client_backend` integration: we need to be able to get the recipient's UA, not just the Orchard component (and same for Sapling). We construct metadata values that carry this data along in the transaction construction process and normally we use it immediately, but with PCZT it needs to be carried along. I guess that this should go into the `proprietary` field; maybe that should be exposed via a mut getter so that it can be manipulated in that way?

nuttycom replied:
We'll also have to add proposal step metadata here. We can't add this metadata directly in the process of adding outputs to the `zcash_primitives` builder; instead, we have to track it separately, then mutate the resulting PCZT to add the metadata so that we can carry it over to where we store the transaction data.

str4d replied:
In earlier PCZT discussions this year I suggested that the transaction proposal should be extractable from the (set of) PCZT resulting from it, but got strong enough pushback against it that I dropped the idea. So yes, if you now do actually need to represent parts of the proposal in the PCZTs, you will need to do so via the `proprietary` field.

I think we will need to expose some kinds of getters and setters here, in particular for the Updater role (which I haven't implemented because it's so broad) because we need to be able to set `Zip32Derivation` values to inform the Keystone device of which keys to use.

nuttycom replied:
It's not actually part of the proposal per se; some of the metadata that we need to carry along is stuff that's derived from wallet state in the process of transaction construction. So I think that the decision to not make a `Proposal` extractable was actually correct, because it would have been incomplete with respect to what we need anyway.

daira commented on pczt/src/lib.rs:
```diff
@@ -1 +1,129 @@
+//!     - A source of randomness.
+//! - Combiner (anyone can execute)
+//!   - Combines several PCZTs that represent the same transaction into a single PCZT.
+//!   - Because we aren't storing the partial transaction in network format, we need to
+//!     carefully define equality for PCZTs.
+//!     - If we say "pczt.global must be identical" then:
+//!       - If we add spends or outputs in series, we should always update bsk when adding
+//!         spends or outputs, even if rcv is present.
+//!       - If we add spends or outputs in parallel and then combine, we must _never_ update
+//!         bsk, and then update it when we prepare for signature creation.
+//!       We can't control which happens, ergo we need an IO Finalizer step.
```
```suggestion
```
(This is rationale that is no longer useful here.)

daira commented on pczt/src/lib.rs:
```diff
@@ -1 +1,129 @@
+impl Pczt {
+    /// Parses a PCZT from its encoding.
+    pub fn parse(bytes: &[u8]) -> Result<Self, ParseError> {
+        if bytes.len() < 8 {
+            return Err(ParseError::TooShort);
+        }
+        if &bytes[..4] != MAGIC_BYTES {
+            return Err(ParseError::NotPczt);
+        }
+        let version = u32::from_le_bytes(bytes[4..8].try_into().unwrap());
+        if version != PCZT_VERSION_1 {
+            return Err(ParseError::UnknownVersion(version));
+        }
+
+        // This is a v1 PCZT.
+        postcard::from_bytes(&bytes[8..]).map_err(ParseError::Postcard)
```
File a ticket for a test that exercises error cases.

str4d replied:
I'm adding the latter to this PR.

daira commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,524 @@
+    ///
+    /// - This is set by the Constructor.
+    /// - This is required by the Prover.
+    ///
+    /// TODO: This could be merged with `rseed` into a tuple. `recipient` and `value` are
+    /// separate because they might need to be independently redacted. (For which role?)
```
```suggestion
```

daira commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,524 @@
+    /// Merges this bundle with another.
+    ///
+    /// Returns `None` if the bundles have conflicting data.
+    pub(crate) fn merge(mut self, other: Self) -> Option<Self> {
```
File a ticket to add a test of the combiner.

str4d replied:
I'll add it to this PR by altering one of the end-to-end tests.

daira replied:
This still appears not to be covered.

str4d replied:
It is covered by `end_to_end::sapling_to_orchard`.

nuttycom commented on zcash_primitives/src/transaction/builder.rs:
```diff
@@ -272,6 +273,30 @@ impl BuildResult {
+/// The components of a PCZT.
+#[derive(Debug)]
+pub struct PcztParts<P: Parameters> {
+    pub params: P,
```
Why does the `PcztParts` need to carry the parameters along with it? This struct is created by `build_for_pczt`, and the parameters are those that were originally provided to the builder.

Given that a builder effectively a curried function, you're just giving back a value that the caller originally passed to you.

str4d replied:
We need to obtain `coin_type` when setting up the PCZT, and given that we already have a pattern of passing around parameters, this seemed like the most type-safe way to do so.

daira commented on pczt/src/roles/creator/mod.rs:
```diff
@@ -0,0 +1,161 @@
+                expiry_height: self.expiry_height,
+                coin_type: self.coin_type,
+                // Spends and outputs modifiable, no SIGHASH_SINGLE.
+                tx_modifiable: 0b0000_0011,
```
Ideally this should use an "or" of symbolic constants, not a hard-coded value. Non-blocking.

daira commented on pczt/src/roles/creator/mod.rs:
```diff
@@ -0,0 +1,161 @@
+                .iter()
+                .any(|input| input.sighash_type() == &SighashType::SINGLE)
+        }) {
+            tx_modifiable |= 0b0000_0100;
```
This should also use a symbolic constant. Non-blocking.

str4d commented:
[Force-pushed](https://github.com/zcash/librustzcash/compare/1e72102cbe74bdcb1a6ffa330aa8c17600c72259..3b1ff142402849058bd0b1a3ace6b28e33164f0a) to fix clippy lints and address review comments.

str4d commented:
[Force-pushed](https://github.com/zcash/librustzcash/compare/3b1ff142402849058bd0b1a3ace6b28e33164f0a..4e9f728596170fc9105f12e95925e18c91bb9894) to expose some error types.

daira commented on pczt/src/common.rs:
```diff
@@ -0,0 +1,200 @@
+
+pub(crate) const FLAG_INPUTS_MODIFIABLE: u8 = 0b0000_0001;
+pub(crate) const FLAG_OUTPUTS_MODIFIABLE: u8 = 0b0000_0010;
+pub(crate) const FLAG_HAS_SIGHASH_SINGLE: u8 = 0b0000_0100;
```
```suggestion
#[allow(dead_code)]
pub(crate) const FLAG_HAS_SIGHASH_SINGLE: u8 = 0b0000_0100;
```

daira commented on pczt/src/roles/signer/mod.rs:
```diff
@@ -0,0 +1,302 @@
+            self.global.tx_modifiable &= !0b0000_0010;
+        }
+        if [SighashType::SINGLE, SighashType::SINGLE].contains(input.sighash_type()) {
+            self.global.tx_modifiable |= 0b0000_0100;
```
```suggestion
            self.global.tx_modifiable |= FLAG_HAS_SIGHASH_SINGLE;
```

daira commented on pczt/src/common.rs:
```diff
@@ -0,0 +1,200 @@
+            self.tx_modifiable &= !0b0000_0010;
+        }
+        // - Bit 2 merges towards `true`.
+        if (tx_modifiable & 0b0000_0100) != 0 {
```
```suggestion
        if (tx_modifiable & FLAG_HAS_SIGHASH_SINGLE) != 0 {
```

daira commented on pczt/src/common.rs:
```diff
@@ -0,0 +1,200 @@
+            self.tx_modifiable &= !0b0000_0010;
+        }
+        // - Bit 2 merges towards `true`.
+        if (tx_modifiable & 0b0000_0100) != 0 {
+            self.tx_modifiable |= 0b0000_0100;
```
```suggestion
        if (tx_modifiable & FLAG_HAS_SIGHASH_SINGLE) != 0 {
            self.tx_modifiable |= FLAG_HAS_SIGHASH_SINGLE;
```

daira commented on pczt/src/roles/signer/mod.rs:
```diff
@@ -0,0 +1,302 @@
+            self.global.tx_modifiable &= !0b0000_0001;
+        }
+        if ![SighashType::NONE, SighashType::NONE_ANYONECANPAY].contains(input.sighash_type()) {
+            self.global.tx_modifiable &= !0b0000_0010;
```
```suggestion
            self.global.tx_modifiable &= !FLAG_OUTPUTS_MODIFIABLE;
```

daira commented on pczt/src/common.rs:
```diff
@@ -0,0 +1,200 @@
+
+        // `tx_modifiable` is explicitly a bitmap; merge it bit-by-bit.
+        // - Bit 0 and Bit 1 merge towards `false`.
+        if (tx_modifiable & 0b0000_0001) == 0 {
+            self.tx_modifiable &= !0b0000_0001;
```
```suggestion
        if (tx_modifiable & FLAG_INPUTS_MODIFIABLE) == 0 {
            self.tx_modifiable &= !FLAG_INPUTS_MODIFIABLE;
```

daira commented on pczt/src/roles/signer/mod.rs:
```diff
@@ -0,0 +1,302 @@
+
+        // Update transaction modifiability.
+        if input.sighash_type().encode() & SIGHASH_ANYONECANPAY == 0 {
+            self.global.tx_modifiable &= !0b0000_0001;
```
```suggestion
            self.global.tx_modifiable &= !FLAG_INPUTS_MODIFIABLE;
```

daira commented on pczt/src/roles/signer/mod.rs:
```diff
@@ -0,0 +1,302 @@
+        if input.sighash_type().encode() & SIGHASH_ANYONECANPAY == 0 {
+            self.global.tx_modifiable &= !0b0000_0001;
+        }
+        if ![SighashType::NONE, SighashType::NONE_ANYONECANPAY].contains(input.sighash_type()) {
```
Would this be clearer?
```suggestion
        if (input.sighash_type() & !SIGHASH_ANYONECANPAY) != SighashType::NONE {
```

daira commented on pczt/Cargo.toml:
```diff
@@ -11,3 +11,61 @@ license.workspace = true
+transparent = ["dep:secp256k1", "dep:zcash_primitives", "dep:zcash_protocol"]
+zcp-builder = ["dep:zcash_primitives"]
+io-finalizer = ["orchard", "sapling"]
+prover = ["dep:rand_core", "sapling?/temporary-zcashd"]
```
This was necessary before due to a use of `SpendValidatingKey::temporary_zcash_from_bytes`, but that appears to have been removed, and this branch now compiles for me without the `temporary-zcashd` feature dependency (using `--all-targets --all-features`).
```suggestion
prover = ["dep:rand_core"]
```

daira replied:
Still relevant.

daira commented on pczt/src/lib.rs:
```diff
@@ -1 +1,122 @@
+    /// The bytes do not contain a PCZT.
+    NotPczt,
+    /// The PCZT encoding was invalid.
+    Postcard(postcard::Error),
```
```suggestion
    Invalid(postcard::Error),
```
The meaning of the error doesn't depend on Postcard, even if the wrapped error type does. A client that only matched on `Invalid(_)`, for example, would not be depending on the encoding being Postcard (or implemented using any particular Postcard library).

daira commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,536 @@
+    /// The Orchard anchor for this transaction.
+    ///
+    /// Set by the Creator.
+    pub(crate) anchor: [u8; 32],
```
Why isn't the anchor optional? It only needs to be available to the Prover, I think, and there could be situations where we would want to set it after creation.

daira commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,536 @@
+
+    /// The Orchard anchor for this transaction.
+    ///
+    /// Set by the Creator.
+    pub(crate) anchor: [u8; 32],
```
Why isn't the anchor optional? It only needs to be available to the Prover, I think, and there could be situations where we would want to set it after creation. In particular, if the length of time needed for construction of a PCZT up to a certain point is significant relative to the block time, then we would want to wait until after that point to set the anchor, so that we can use an anchor from a recent block.
```suggestion
    /// Set by the Creator or Constructor.
    pub(crate) anchor: Option<[u8; 32]>,
```

nuttycom replied:
The Constructor needs to know the anchors that have been chosen because the choice of anchors places a restriction on what spends can be contributed, so they must be set by the Creator. Also, since anchor choice has implications for privacy, all of the Constructors need to be able to agree on the anchors; as such, I don't think that they can be optional. The anchors are public in the finished transaction, so there's never any need to redact them.

daira replied:
That's a good argument. I withdraw the suggestion (also for `expiry_height`, since my rationale depended on all of `expiry_height` and the Sapling and Orchard anchors being modifiable).

daira commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,536 @@
+            }
+        }
+
+        if self.anchor != anchor {
```
```suggestion
        if !merge_optional(&mut self.anchor, anchor) {
```
(if `anchor` is made optional).

daira commented on pczt/src/lib.rs:
```diff
@@ -1 +1,122 @@
+//!     then receive back the PCZT with private data stripped and proof added.
+//! - Signer (capability holders can contribute)
+//!   - Needs the spend authorization randomizers to create signatures.
+//!   - Needs sufficient information to verify that the proof is over the correct data.
```
```suggestion
//!   - Needs sufficient information to verify that the proof is over the correct data,
```

daira commented on pczt/src/roles/signer/mod.rs:
```diff
@@ -0,0 +1,302 @@
+
+            let nk = proof_generation_key.to_viewing_key().nk;
+
+            let merkle_path = spend.witness().as_ref().ok_or(Error::MissingWitness)?;
+
+            if &note.nf(&nk, merkle_path.position().into()) != tx_spend.nullifier() {
```
I thought that this might be imposing an unnecessary constraint by requiring the witness of the spent note rather than just the position. But actually the constraint is necessary, because the sighash for Sapling spend auth signatures is over the anchor [even in v5 transactions](https://zips.z.cash/zip-0244#t-3a-ii-sapling-spends-noncompact-digest). And so even if the anchor is made optional [as I suggested](https://github.com/zcash/librustzcash/pull/1577/files#r1873734850), we cannot update it after making signatures. (The sighash for Orchard spend auth signatures is [also over the anchor](https://zips.z.cash/zip-0244#t-4-orchard-digest).)

daira commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,536 @@
+            actions,
+            flags: bundle.flags().to_byte(),
+            value_sum,
+            anchor: bundle.anchor().to_bytes(),
```
```suggestion
            anchor: Some(bundle.anchor().to_bytes()),
```

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,526 @@
+
+    /// The Sapling anchor for this transaction.
+    ///
+    /// Set by the Creator.
+    pub(crate) anchor: [u8; 32],
```
Similar reasoning to Orchard:
```suggestion
    /// Set by the Creator or Constructor.
    pub(crate) anchor: Option<[u8; 32]>,
```

daira commented on pczt/src/common.rs:
```diff
@@ -0,0 +1,200 @@
+    /// - If omitted, the fallback locktime is assumed to be 0.
+    pub(crate) fallback_lock_time: Option<u32>,
+
+    pub(crate) expiry_height: u32,
```
I think the expiry height should be optional. If the length of time needed for construction of a PCZT up to a certain point is significant relative to the block time, then we would want to wait until after that point to set the expiry height.

(We should make the same decision about `expiry_height`, `sapling_anchor`, and `orchard_anchor`, since the same argument about potentially long-lived PCZTs applies to all of them.)

nuttycom replied:
The expiry height has implications for how long the inputs that were contributed should be considered to be "locked up" for until they can be available for spending again, if the transaction is not completed. As such, the expiry height needs to be known to and agreed upon by the Constructors, and should just be chosen such that there's enough time to complete the transaction at the outset. I feel less strongly about this than I do about the anchors, which I think _must_ be set by the `Creator`; one downside of requiring the expiry height to be set could be that some users then instead choose a zero expiry height in order to "allow more time" for transaction construction. But if they want to do that, they can just set an expiry height that's far in the future anyway, and at least then there's _some_ bound. Keep this required.

daira commented on pczt/src/common.rs:
```diff
@@ -0,0 +1,200 @@
+            version_group_id: 0,
+            consensus_branch_id: 0,
+            fallback_lock_time: None,
+            expiry_height: 0,
```
```suggestion
            expiry_height: None,
```

daira commented on pczt/src/roles/creator/mod.rs:
```diff
@@ -0,0 +1,164 @@
+    version_group_id: u32,
+    consensus_branch_id: u32,
+    fallback_lock_time: Option<u32>,
+    expiry_height: u32,
```
```suggestion
    expiry_height: Option<u32>,
```

daira commented on pczt/src/roles/creator/mod.rs:
```diff
@@ -0,0 +1,164 @@
+    expiry_height: u32,
+    coin_type: u32,
+    orchard_flags: u8,
+    sapling_anchor: [u8; 32],
+    orchard_anchor: [u8; 32],
```
```suggestion
    sapling_anchor: Option<[u8; 32]>,
    orchard_anchor: Option<[u8; 32]>,
```

daira commented on pczt/src/roles/creator/mod.rs:
```diff
@@ -0,0 +1,164 @@
+impl Creator {
+    pub fn new(
+        consensus_branch_id: u32,
+        expiry_height: u32,
+        coin_type: u32,
+        sapling_anchor: [u8; 32],
+        orchard_anchor: [u8; 32],
```
```suggestion
        expiry_height: Option<u32>,
        coin_type: u32,
        sapling_anchor: Option<[u8; 32]>,
        orchard_anchor: Option<[u8; 32]>,
```

daira commented on pczt/src/roles/creator/mod.rs:
```diff
@@ -0,0 +1,164 @@
+                version_group_id: parts.version.version_group_id(),
+                consensus_branch_id: parts.consensus_branch_id.into(),
+                fallback_lock_time: Some(parts.lock_time),
+                expiry_height: parts.expiry_height.into(),
```
```suggestion
                expiry_height: parts.expiry_height.map(|h| h.into()),
```

daira commented on zcash_primitives/src/transaction/builder.rs:
```diff
@@ -272,6 +273,30 @@ impl BuildResult {
+    pub version: TxVersion,
+    pub consensus_branch_id: BranchId,
+    pub lock_time: u32,
+    pub expiry_height: BlockHeight,
```
```suggestion
    pub expiry_height: Option<BlockHeight>,
```

daira commented on pczt/src/roles/signer/mod.rs:
```diff
@@ -0,0 +1,302 @@
+        consensus_branch_id,
+        determine_lock_time(global, transparent.inputs())
+            .map_err(|_| Error::IncompatibleLockTimes)?,
+        global.expiry_height.into(),
```
```suggestion
        global.expiry_height.map(|h| h.into()),
```

daira commented on pczt/src/roles/tx_extractor/mod.rs:
```diff
@@ -0,0 +1,257 @@
+            version,
+            consensus_branch_id,
+            lock_time,
+            pczt.global.expiry_height.into(),
```
```suggestion
            pczt.global.expiry_height.map(|h| h.into()),
```

daira commented on pczt/src/roles/creator/mod.rs:
```diff
@@ -0,0 +1,164 @@
+                    spends: vec![],
+                    outputs: vec![],
+                    value_sum: 0,
+                    anchor: sapling::Anchor::empty_tree().to_bytes(),
```
```suggestion
                    anchor: None,
```

daira commented on pczt/src/roles/creator/mod.rs:
```diff
@@ -0,0 +1,164 @@
+                    actions: vec![],
+                    flags: ORCHARD_SPENDS_AND_OUTPUTS_ENABLED,
+                    value_sum: (0, true),
+                    anchor: orchard::Anchor::empty_tree().to_bytes(),
```
```suggestion
                    anchor: None,
```

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,526 @@
+            }
+        }
+
+        if self.anchor != anchor {
```
```suggestion
        if !merge_optional(self.anchor, anchor) {
```

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,526 @@
+            spends,
+            outputs,
+            value_sum: bundle.value_sum().to_raw(),
+            anchor: bundle.anchor().to_bytes(),
```
```suggestion
            anchor: bundle.anchor().map(|anchor| anchor.to_bytes()),
```

daira commented on pczt/src/roles/io_finalizer/mod.rs:
```diff
@@ -0,0 +1,105 @@
+        // After shielded IO finalization, the inputs and outputs cannot be modified
+        // because dummy spends will have been signed.
+        if has_shielded_spends || has_shielded_outputs {
+            global.tx_modifiable &= !0b0000_0011;
```
```suggestion
            global.tx_modifiable &= !(FLAG_INPUTS_MODIFIABLE | FLAG_OUTPUTS_MODIFIABLE);
```

daira commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,536 @@
+    /// Encoded as a `Vec<u8>` because its length depends on the transaction version.
+    pub(crate) out_ciphertext: Vec<u8>,
+
+    /// The address that will receive the output.
```
```suggestion
    /// The [Orchard raw payment address](https://zips.z.cash/protocol/protocol-dark.pdf#orchardpaymentaddrencoding)
    /// that will receive the output.
```

daira commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,536 @@
+    #[serde_as(as = "Option<[_; 64]>")]
+    pub(crate) spend_auth_sig: Option<[u8; 64]>,
+
+    /// The address that received the note being spent.
```
```suggestion
    /// The [Orchard raw payment address](https://zips.z.cash/protocol/protocol-dark.pdf#orchardpaymentaddrencoding)
    /// that received the note being spent.
```

daira commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,536 @@
+
+    /// The address that received the note being spent.
+    ///
+    /// - This is set by the Constructor (or Updater?).
```
```suggestion
    /// - This is set by the Constructor or Updater.
```

str4d replied:
It's actually set by the Constructor, because the Constructor necessarily knows it. I'll update the text to fix.

daira commented on pczt/src/roles.rs:
```diff
@@ -0,0 +1,64 @@
+            tx_extractor::{self, TransactionExtractor},
+        };
+
+        let pczt = Creator::new(BranchId::Nu6.into(), 10_000_000, 133, [0; 32], [0; 32]).build();
```
```suggestion
        let pczt = Creator::new(BranchId::Nu6.into(), Some(10_000_000), 133, Some([0; 32]), Some([0; 32])).build();
```

daira commented on pczt/src/roles.rs:
```diff
@@ -0,0 +1,64 @@
+            io_finalizer::{self, IoFinalizer},
+        };
+
+        let pczt = Creator::new(BranchId::Nu6.into(), 10_000_000, 133, [0; 32], [0; 32]).build();
```
```suggestion
        let pczt = Creator::new(BranchId::Nu6.into(), Some(10_000_000), 133, Some([0; 32]), Some([0; 32])).build();
```

daira commented on pczt/src/roles/combiner/mod.rs:
```diff
@@ -0,0 +1,87 @@
+    rhs: BTreeMap<K, V>,
+) -> bool {
+    for (key, rhs_value) in rhs.into_iter() {
+        if let Some(lhs_value) = lhs.get_mut(&key) {
```
```suggestion
        if let Some(lhs_value) = lhs.get(&key) {
```

daira commented on pczt/src/roles/signer/mod.rs:
```diff
@@ -0,0 +1,302 @@
+        if input.sighash_type().encode() & SIGHASH_ANYONECANPAY == 0 {
+            self.global.tx_modifiable &= !0b0000_0001;
+        }
+        if ![SighashType::NONE, SighashType::NONE_ANYONECANPAY].contains(input.sighash_type()) {
```
If (as above) it is intended that the modifiability flags only prevent modifications that are definitely incorrect, then the outputs should only be made non-modifiable for `SighashType::{ALL, ALL_ANYONECANPAY}`. In the case of `SighashType::{SINGLE, SINGLE_ANYONECANPAY}`, other outputs than the one corresponding to the input being signed can still be modified.
```suggestion
        if (input.sighash_type() & !SIGHASH_ANYONECANPAY) == SighashType::ALL {
```

Reference for the semantics of sighash types: https://developer.bitcoin.org/devguide/transactions.html#signature-hash-types

str4d replied:
The semantics of `tx_modifiable` come directly from [BIP 370](https://github.com/bitcoin/bips/blob/master/bip-0370.mediawiki#signer) (formatting added for clarity):
> For PSBTv2s, a signer must update the PSBT_GLOBAL_TX_MODIFIABLE field after signing inputs so that it accurately reflects the state of the PSBT.
> - If the Signer added a signature that does not use SIGHASH_ANYONECANPAY, the Input Modifiable flag must be set to False.
> - If the Signer added a signature that does not use SIGHASH_NONE, the Outputs Modifiable flag must be set to False.
> - If the Signer added a signature that uses SIGHASH_SINGLE, the Has SIGHASH_SINGLE flag must be set to True. 

Technically someone can add an output after an existing SIGHASH_SINGLE-constrained one, but given that we want to match BIP 370 semantics where possible for ease of usage alongside PSBTs, I'm fine with this being more restrictive.

daira commented on pczt/src/roles/signer/mod.rs:
```diff
@@ -0,0 +1,302 @@
+        if ![SighashType::NONE, SighashType::NONE_ANYONECANPAY].contains(input.sighash_type()) {
+            self.global.tx_modifiable &= !0b0000_0010;
+        }
+        if [SighashType::SINGLE, SighashType::SINGLE].contains(input.sighash_type()) {
```
This was presumably supposed to be `[SighashType::SINGLE, SighashType::SINGLE_ANYONECANPAY]`. I think it's clearer like this though:
```suggestion
        if (input.sighash_type() & !SIGHASH_ANYONECANPAY) == SighashType::SINGLE {
```

daira commented on pczt/src/roles/io_finalizer/mod.rs:
```diff
@@ -0,0 +1,105 @@
+        let tx_data = pczt_to_tx_data(&global, &transparent, &sapling, &orchard)?;
+        let txid_parts = tx_data.digest(TxIdDigester);
+
+        // TODO: Pick sighash based on tx version.
```
Is this going to be fixed in this PR? It seems fairly straightforward.

str4d replied:
It isn't straightforward because with v4 transactions the Sapling proofs need to be created before we can derive a sighash. I'm not actually sure if the current PCZT format even supports v4 transactions (I have not tried creating Sapling proofs before running the IO Finalizer). For now, we will only support v5 transactions, and then we can revisit this when we get v6 (as that will also necessitate a PCZT v2 encoding).

daira replied:
Please document this limitation. Also, in that case should we make it impossible for now to create a PCZT with `tx_version < 5`? The current behaviour is just to use the wrong sighash calculation, I think, which seems like a significant footgun.

str4d replied:
Rather than making it impossible, I'll gate the roles that require sighashing on the version being compatible.

daira commented on pczt/src/roles/tx_extractor/mod.rs:
```diff
@@ -0,0 +1,257 @@
+            .max()
+            .expect("iterator is non-empty because have_required_lock_time is true")),
+        // If a PSBT has both types of locktimes possible because one or more inputs
+        // specify both `required_time_lock_time` and `required_height_lock_time`, then
```
```suggestion
        // specify both `required_time_lock_time` and `required_height_lock_time`, then a
```

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,526 @@
+    #[serde_as(as = "Option<[_; GROTH_PROOF_SIZE]>")]
+    pub(crate) zkproof: Option<[u8; GROTH_PROOF_SIZE]>,
+
+    /// The address that will receive the output.
```
```suggestion
    /// The [Sapling payment address](https://zips.z.cash/protocol/protocol.pdf#saplingpaymentaddrencoding)
    /// that will receive the output.
```

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,526 @@
+    #[serde_as(as = "Option<[_; 64]>")]
+    pub(crate) spend_auth_sig: Option<[u8; 64]>,
+
+    /// The address that received the note being spent.
```
```suggestion
    /// The [raw encoding](https://zips.z.cash/protocol/protocol.pdf#saplingpaymentaddrencoding)
    /// of the Sapling payment address that received the note being spent.
```

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,526 @@
+    /// The note commitment randomness.
+    ///
+    /// - This is set by the Constructor. It MUST NOT be set if the note has an `rseed`
+    ///   (i.e. was created after [ZIP 212]).
```
```suggestion
    ///   (i.e. was created after [ZIP 212] activation).
```

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,526 @@
+    /// The seed randomness for the note being spent.
+    ///
+    /// - This is set by the Constructor. It MUST NOT be set if the note has no `rseed`
+    ///   (i.e. was created before [ZIP 212]).
```
```suggestion
    ///   (i.e. was created before [ZIP 212] activation).
```

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,526 @@
+
+    /// The address that received the note being spent.
+    ///
+    /// - This is set by the Constructor (or Updater?).
```
```suggestion
    /// - This is set by the Constructor or Updater.
```

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,526 @@
+    /// The value commitment randomness.
+    ///
+    /// - This is set by the Constructor.
+    /// - The IO Finalizer compresses it into the bsk.
```
```suggestion
    /// - The IO Finalizer compresses it into `bsk`.
```

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,526 @@
+    /// - This is chosen by the Constructor.
+    /// - This is required by the Signer for creating `spend_auth_sig`, and may be used to
+    ///   validate `rk`.
+    /// - After`zkproof` / `spend_auth_sig` has been set, this can be redacted.
```
```suggestion
    /// - After `zkproof` / `spend_auth_sig` has been set, this can be redacted.
```

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,526 @@
+    ///
+    /// Encoded as a `Vec<u8>` because its length depends on the transaction version.
+    ///
+    /// Once we have memo bundles, we will be able to set memos independently of Outputs.
+    /// For now, the Constructor sets both at the same time.
```
```suggestion
    /// Once we have [memo bundles], we will be able to set memos independently of Outputs.
    /// For now, the Constructor sets both at the same time.
    ///
    /// [memo bundles]: https://zips.z.cash/zip-0231
```

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,526 @@
+    /// The value commitment randomness.
+    ///
+    /// - This is set by the Constructor.
+    /// - The IO Finalizer compresses it into the bsk.
```
```suggestion
    /// - The IO Finalizer compresses it into `bsk`.
```

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,526 @@
+            // construction.
+            (None, None) => {
+                if spends.len() > self.spends.len() {
+                    // TODO: Update `self.value_balance`.
+                    self.spends.extend(spends.drain(self.spends.len()..));
+                }
+                if outputs.len() > self.outputs.len() {
+                    // TODO: Update `self.value_balance`.
```
These seem like important TODOs.

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,526 @@
+            // construction.
+            (None, None) => {
+                if spends.len() > self.spends.len() {
+                    // TODO: Update `self.value_balance`.
+                    self.spends.extend(spends.drain(self.spends.len()..));
+                }
+                if outputs.len() > self.outputs.len() {
+                    // TODO: Update `self.value_balance`.
```
I think these TODOs need to be resolved, otherwise it leaves an incorrect `self.value_balance`. (I see why it is more complicated than in the Orchard case: it may be that one bundle has longer spends, and the other has longer outputs. Maybe we can just return `None` in that case?)

str4d replied:
You are correct; good catch. I'll just return `None` in the case we can't handle without parsed bundles.

daira commented on pczt/src/transparent.rs:
```diff
@@ -0,0 +1,414 @@
+    ///
+    /// - Signers must use this sighash type to produce their signatures. Signers that
+    ///   cannot produce signatures for this sighash type must not provide a signature.
+    /// - Spend Finalizers must fail to finalize inputs which have signatures that do not
+    ///   match this sighash type.
```
```suggestion
    /// - Spend Finalizers must fail to finalize inputs that have signatures not matching
    ///   this sighash type.
```

daira commented on pczt/src/transparent.rs:
```diff
@@ -0,0 +1,414 @@
+    ///
+    /// - The pubkeys should appear in `script_pubkey` or `redeem_script`.
+    /// - Each entry is set by an Updater.
+    /// - Individual entries may be required by a Signer.
```
```suggestion
    /// - Individual entries may be required by a Signer.
    /// - It is not required that the map include entries for all of the used pubkeys.
    ///   In particular, it is not possible to include entries for non-BIP-32 pubkeys.
```

daira commented on zcash_primitives/src/transaction/builder.rs:
```diff
@@ -272,6 +273,30 @@ impl BuildResult {
+}
+
 /// Generates a [`Transaction`] from its inputs and outputs.
 pub struct Builder<'a, P, U: sapling::builder::ProverProgress> {
```
Why `sapling_asks`, but `orchard_saks`?

str4d replied:
"Orchard Spend Authorizing Keys" is where the name came from. These fields go away in the next PR anyway; we can bikeshed there or elsewhere.

daira commented on zcash_primitives/src/transaction/builder.rs:
```diff
@@ -835,6 +860,86 @@ impl<'a, P: consensus::Parameters, U: sapling::builder::ProverProgress> Builder<
+        let balance_after_fees =
+            (self.value_balance()? - fee.into()).ok_or(BalanceError::Underflow)?;
+
+        match balance_after_fees.cmp(&Amount::zero()) {
+            Ordering::Less => {
+                return Err(Error::InsufficientFunds(-balance_after_fees));
+            }
+            Ordering::Greater => {
+                return Err(Error::ChangeRequired(balance_after_fees));
+            }
+            Ordering::Equal => (),
+        };
```
```suggestion
        if balance_after_fees < Amount::zero() {
            return Err(Error::InsufficientFunds(-balance_after_fees));
        } else if balance_after_fees > Amount::zero() {
            return Err(Error::ChangeRequired(balance_after_fees));
        }
```

str4d replied:
Not changing this in this PR because it matches the code in the existing builder function.

daira commented on zcash_primitives/src/transaction/components/transparent/pczt.rs:
```diff
@@ -0,0 +1,217 @@
+    ///
+    /// - The pubkeys should appear in `script_pubkey` or `redeem_script`.
+    /// - Each entry is set by an Updater.
+    /// - Individual entries may be required by a Signer.
```
```suggestion
    /// - Individual entries may be required by a Signer.
    /// - It is not required that the map include entries for all of the used pubkeys.
    ///   In particular, it is not possible to include entries for non-BIP-32 pubkeys.
```

daira commented on zcash_primitives/src/transaction/components/transparent/pczt.rs:
```diff
@@ -0,0 +1,217 @@
+mod updater;
+pub use updater::{InputUpdater, OutputUpdater, Updater, UpdaterError};
+
+mod signer;
```
Should this be behind a feature flag to avoid an unconditional dependency on secp256k1? (I don't understand how it can currently build without "transparent-inputs".)

daira commented on zcash_primitives/src/transaction/components/transparent/pczt.rs:
```diff
@@ -0,0 +1,217 @@
+mod updater;
+pub use updater::{InputUpdater, OutputUpdater, Updater, UpdaterError};
+
+mod signer;
+pub use signer::SignerError;
```
Should this be behind a feature flag to avoid an unconditional dependency on secp256k1? It doesn't currently build without "transparent-inputs".
```suggestion
#[cfg(feature = "transparent-inputs")]
mod signer;
#[cfg(feature = "transparent-inputs")]
pub use signer::SignerError;
```

daira commented on zcash_primitives/src/transaction/components/transparent/pczt.rs:
```diff
@@ -0,0 +1,217 @@
+mod parse;
+pub use parse::ParseError;
+
+mod updater;
+pub use updater::{InputUpdater, OutputUpdater, Updater, UpdaterError};
+
+mod signer;
+pub use signer::SignerError;
+
+mod spend_finalizer;
+pub use spend_finalizer::SpendFinalizerError;
```
Should these be behind "transparent-inputs" to avoid an unconditional dependency on secp256k1 and ripemd? They don't currently build without it.
```suggestion
#[cfg(feature = "transparent-inputs")]
mod updater;
#[cfg(feature = "transparent-inputs")]
pub use updater::{InputUpdater, OutputUpdater, Updater, UpdaterError};

#[cfg(feature = "transparent-inputs")]
mod signer;
#[cfg(feature = "transparent-inputs")]
pub use signer::SignerError;

#[cfg(feature = "transparent-inputs")]
mod spend_finalizer;
#[cfg(feature = "transparent-inputs")]
pub use spend_finalizer::SpendFinalizerError;
```

str4d replied:
I'm going with option 1. The `transparent-inputs` feature flag is for removing the `secp256k1` dependency, and transparent signing has a hard dependency on it (finer-grained would mean there's nothing actually left in the `signer` module).

daira commented on zcash_primitives/src/transaction/components/transparent/builder.rs:
```diff
@@ -9,7 +10,7 @@ use crate::{
             transparent::{self, Authorization, Authorized, Bundle, TxIn, TxOut},
         },
-        sighash::TransparentAuthorizingContext,
+        sighash::{SighashType, TransparentAuthorizingContext},
```
To avoid an unused input warning, `SighashType` should be imported only if "transparent-inputs" is enabled.

daira commented on zcash_primitives/src/transaction/sighash.rs:
```diff
@@ -20,6 +20,39 @@ pub const SIGHASH_SINGLE: u8 = 0x03;
+    ///
+    /// [ZIP 244]: https://zips.z.cash/zip-0244#s-2a-hash-type
+    pub fn parse(hash_type: u8) -> Option<Self> {
+        match hash_type & 0x7f {
```
```suggestion
        match hash_type & !SIGHASH_ANYONECANPAY {
```

daira commented on zcash_proofs/CHANGELOG.md:
Also update the `zcash_primitives` changelog.

daira reviewed (dismissed):
utACK with suggestions.

str4d commented:
[Force-pushed](https://github.com/zcash/librustzcash/compare/4e9f728596170fc9105f12e95925e18c91bb9894..3934b8e5562336e8fa50c431a8b6afc8605efc08) to bring in getter-related changes to `pczt` from #1635, and address most of @daira's comments.

str4d commented on pczt/src/roles/signer/mod.rs:
```diff
@@ -0,0 +1,302 @@
+            .map_err(Error::TransparentSign)?;
+
+        // Update transaction modifiability.
+        if input.sighash_type().encode() & SIGHASH_ANYONECANPAY == 0 {
```
You've read this backwards: we are making all inputs non-modifiable if a signature is added that does not have `ANYONECANPAY` set, i.e. precisely the case you argue for.

daira replied:
I'm confused because signing using *any* sighash algorithm makes at least one of the inputs non-modifiable. I.e. `self.global.tx_modifiable &= !FLAG_INPUTS_MODIFIABLE;` can be unconditional if that's the intent. That's why I thought the intent was only to prevent definitely incorrect cases.

daira replied:
The confusion was that I thought an input/output could be modified in-place.

daira replied:
Actually, an input/output *can* be modified in place because of the `*_mut` accessors, so I'm still confused (as expressed in other comments).

str4d commented:
[Force-pushed](https://github.com/zcash/librustzcash/compare/3934b8e5562336e8fa50c431a8b6afc8605efc08..9b6276bb554cec59b394e552d29e1b44ce4b0ef9) to address more of @daira's comments.

daira commented on pczt/src/roles/signer/mod.rs:
```diff
@@ -0,0 +1,311 @@
+            .map_err(Error::TransparentSign)?;
+
+        // Update transaction modifiability:
+        // - If the Signer added a signature that does not use SIGHASH_ANYONECANPAY, the
+        //   Input Modifiable flag must be set to False.
+        if input.sighash_type().encode() & SIGHASH_ANYONECANPAY == 0 {
+            self.global.tx_modifiable &= !FLAG_INPUTS_MODIFIABLE;
+        }
```
```suggestion
        // - Signatures for all sighash types sign at least one input, so the Input Modifiable
        //   flag must be set to False.
        self.global.tx_modifiable &= !FLAG_INPUTS_MODIFIABLE;
```

daira commented on pczt/src/roles/creator/mod.rs:
```diff
@@ -0,0 +1,164 @@
+            zcash_primitives::transaction::TxVersion::Zip225 => Some(V5_TX_VERSION),
+        }?;
+
+        // Spends and outputs not modifiable.
```
Please clarify this (here and/or in the field definition) to say that it only applies to additions and removals.

daira commented on pczt/src/roles/signer/mod.rs:
```diff
@@ -0,0 +1,311 @@
+
+        // Update transaction modifiability:
+        // - If the Signer added a signature that does not use SIGHASH_ANYONECANPAY, the
+        //   Input Modifiable flag must be set to False.
```
```suggestion
        //   Input Modifiable flag must be set to False (because all inputs are
        //   signed, not just the one at `index`).
```

daira commented on pczt/src/common.rs:
```diff
@@ -0,0 +1,202 @@
+
+use crate::roles::combiner::merge_map;
+
+pub(crate) const FLAG_INPUTS_MODIFIABLE: u8 = 0b0000_0001;
+pub(crate) const FLAG_OUTPUTS_MODIFIABLE: u8 = 0b0000_0010;
+pub(crate) const FLAG_HAS_SIGHASH_SINGLE: u8 = 0b0000_0100;
```
```suggestion
/// Inputs/spends can be neither added nor removed.
pub(crate) const FLAG_INPUTS_MODIFIABLE: u8 = 0b0000_0001;
/// Outputs can be neither added nor removed.
pub(crate) const FLAG_OUTPUTS_MODIFIABLE: u8 = 0b0000_0010;
/// At least one input was signed using SIGHASH_SINGLE (with or without ANYONECANPAY).
pub(crate) const FLAG_HAS_SIGHASH_SINGLE: u8 = 0b0000_0100;
```

str4d replied:
> The design of these flags seems to suggest that we should be able to interleave adding/removing inputs and signing

If you're making a transparent-only transaction then in theory you could do so. This aspect is carried over from the existing deployed PSBT design, in order to make adoption of PCZTs easier in theory (by reducing the delta).

> One possibility is to have a separate `FLAG_SHIELDED_MODIFIABLE` that is always set when we sign any input.

I'm on board with this; then the existing two flags become simply about transparent inputs and outputs, which further reduces the semantic delta between PSBTs and PCZTs.

daira commented on pczt/src/roles/creator/mod.rs:
```diff
@@ -0,0 +1,164 @@
+            bundle
+                .inputs()
+                .iter()
+                .any(|input| input.sighash_type() == &SighashType::SINGLE)
```
```suggestion
                .any(|input| input.sighash_type() & !SIGHASH_ANYONECANPAY == &SighashType::SINGLE)
```
(will need rustfmt I think).

daira commented on pczt/src/roles/creator/mod.rs:
```diff
@@ -0,0 +1,164 @@
+
+        // Spends and outputs not modifiable.
+        let mut tx_modifiable = 0b0000_0000;
+        // Check if any input is using `SIGHASH_SINGLE`.
```
```suggestion
        // Check if any input is using `SIGHASH_SINGLE` (with or without `ANYONECANPAY`).
```

daira commented on pczt/src/roles/signer/mod.rs:
```diff
@@ -0,0 +1,311 @@
+            self.global.tx_modifiable &= !FLAG_INPUTS_MODIFIABLE;
+        }
+        // - If the Signer added a signature that does not use SIGHASH_NONE, the Outputs
+        //   Modifiable flag must be set to False.
```
```suggestion
        //   Modifiable flag must be set to False. This also applies to SIGHASH_SINGLE
        //   because we could remove and then re-add the output at `index`, which
        //   would not remove the signature.
```

daira commented on pczt/src/transparent.rs:
```diff
@@ -0,0 +1,411 @@
+            mut outputs,
+        } = other;
+
+        // If the other bundle has more inputs or outputs than us, move them over; these
+        // cannot conflict by construction.
+        self.inputs.extend(inputs.drain(self.inputs.len()..));
+        self.outputs.extend(outputs.drain(self.outputs.len()..));
```
This does not check `tx_modifiable`. Is it supposed to?

str4d replied:
Good catch, it should enforce that if `tx_modifiable` on a bundle indicates no modifications, then merging of the other bundle should fail if it would add inputs or outputs (a combine cannot alter or remove inputs or outputs).

daira commented on pczt/src/common.rs:
```diff
@@ -0,0 +1,202 @@
+pub(crate) const FLAG_INPUTS_MODIFIABLE: u8 = 0b0000_0001;
+pub(crate) const FLAG_OUTPUTS_MODIFIABLE: u8 = 0b0000_0010;
+pub(crate) const FLAG_HAS_SIGHASH_SINGLE: u8 = 0b0000_0100;
+
```
Shielded inputs and outputs are always signed; that [isn't dependent on the sighash flags or whether we're signing a transparent input](https://zips.z.cash/zip-0244#s-3-sapling-digest). Therefore I think we need a separate `FLAG_SHIELDED_MODIFIABLE` that is *always* set when we sign any input.

daira commented on pczt/src/roles/creator/mod.rs:
```diff
@@ -0,0 +1,164 @@
+                fallback_lock_time: self.fallback_lock_time,
+                expiry_height: self.expiry_height,
+                coin_type: self.coin_type,
+                // Spends and outputs modifiable, no SIGHASH_SINGLE.
+                tx_modifiable: FLAG_INPUTS_MODIFIABLE | FLAG_OUTPUTS_MODIFIABLE,
```
```suggestion
                // Spends, outputs, and shielded components modifiable, no SIGHASH_SINGLE.
                tx_modifiable: INITIAL_TX_MODIFIABLE,
```

daira commented on pczt/src/common.rs:
```diff
@@ -0,0 +1,202 @@
+
+    /// A bitfield for various transaction modification flags.
+    ///
+    /// - Bit 0 is the Inputs Modifiable Flag and indicates whether inputs can be modified.
+    ///   - This is set to `true` by the Creator.
+    ///   - This is checked by the Constructor before adding inputs, and may be set to
+    ///     `false` by the Constructor.
+    ///   - This is set to `false` by the IO Finalizer if there are shielded spends or
+    ///     outputs.
+    ///   - This is set to `false` by a Signer that adds a signature that does not use
+    ///     `SIGHASH_ANYONECANPAY`.
+    ///   - The Combiner merges this bit towards `false`.
+    /// - Bit 1 is the Outputs Modifiable Flag and indicates whether outputs can be
+    ///   modified.
+    ///   - This is set to `true` by the Creator.
+    ///   - This is checked by the Constructor before adding outputs, and may be set to
+    ///     `false` by the Constructor.
+    ///   - This is set to `false` by the IO Finalizer if there are shielded spends or
+    ///     outputs.
+    ///   - This is set to `false` by a Signer that adds a signature that does not use
+    ///     `SIGHASH_NONE`.
+    ///   - The Combiner merges this bit towards `false`.
+    /// - Bit 2 is the Has `SIGHASH_SINGLE` flag and indicates whether the transaction has
+    ///   a `SIGHASH_SINGLE` signature who's input and output pairing must be preserved.
+    ///   - This is set to `false` by the Creator.
+    ///   - This is updated by a Constructor.
+    ///   - This is set to `true` by a Signer that adds a signature that uses
+    ///     `SIGHASH_SINGLE`.
+    ///   - This essentially indicates that the Constructor must iterate the transparent
+    ///     inputs to determine whether and how to add a transparent input.
+    ///   - The Combiner merges this bit towards `true`.
+    /// - Bits 3-7 must be 0.
```
All of this would obviously need to be changed if adding `FLAG_SHIELDED_MODIFIABLE`, but I'll wait to see if I'm missing something first.

str4d replied:
I made changes to this documentation as part of defining Bit 7.

daira commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,535 @@
+            // construction.
+            (None, None) => {
+                if actions.len() > self.actions.len() {
+                    self.actions.extend(actions.drain(self.actions.len()..));
```
This does not check whether we have any signatures. If we do then modifying `self.actions` would invalidate them.

str4d replied:
Checking the `tx_modifiable` flags, and relying on the Signer to update them correctly, should be sufficient here. (If they don't, the only effect is an identifiable DoS because the final tx extraction will fail at signature validation.)

daira commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,535 @@
+                return None;
+            }
+
+            if !(merge_optional(&mut lhs.spend.spend_auth_sig, spend_auth_sig)
+                && merge_optional(&mut lhs.spend.recipient, recipient)
+                && merge_optional(&mut lhs.spend.value, value)
+                && merge_optional(&mut lhs.spend.rho, rho)
+                && merge_optional(&mut lhs.spend.rseed, rseed)
+                && merge_optional(&mut lhs.spend.fvk, fvk)
+                && merge_optional(&mut lhs.spend.witness, witness)
+                && merge_optional(&mut lhs.spend.alpha, alpha)
+                && merge_optional(&mut lhs.spend.zip32_derivation, spend_zip32_derivation)
+                && merge_optional(&mut lhs.spend.dummy_sk, dummy_sk)
+                && merge_map(&mut lhs.spend.proprietary, spend_proprietary)
+                && merge_optional(&mut lhs.output.recipient, output_recipient)
+                && merge_optional(&mut lhs.output.value, output_value)
+                && merge_optional(&mut lhs.output.rseed, output_rseed)
+                && merge_optional(&mut lhs.output.ock, ock)
+                && merge_optional(&mut lhs.output.zip32_derivation, output_zip32_derivation)
+                && merge_map(&mut lhs.output.proprietary, output_proprietary)
+                && merge_optional(&mut lhs.rcv, rcv))
```
None of the effecting fields (here, the fields directly represented in the transaction) can be merged in a way that modifies them if we have any signatures.

daira commented on pczt/src/roles/signer/mod.rs:
```diff
@@ -0,0 +1,311 @@
+                return Err(Error::InvalidNullifier);
+            }
+        }
+
```
```suggestion

        // The sighash for a Sapling spend depends on all effecting data.
        self.tx_modifiable &= !(FLAG_INPUTS_MODIFIABLE | FLAG_OUTPUTS_MODIFIABLE);

```
(If `spend.sign` is guaranteed to be non-mutating when it fails, this could be after the call.)

str4d replied:
I originally didn't include these because I believed that shielded signatures couldn't be applied before the IO Finalizer ran, and that sets the necessary flags. However, looking at it again I think that these could potentially be reordered, so I'll ensure the flags get set regardless for shielded signatures.

daira commented on pczt/src/roles/signer/mod.rs:
```diff
@@ -0,0 +1,311 @@
+                return Err(Error::InvalidNullifier);
+            }
+        }
+
```
```suggestion

        // The sighash for an Orchard spend depends on all effecting data.
        self.tx_modifiable &= !(FLAG_INPUTS_MODIFIABLE | FLAG_OUTPUTS_MODIFIABLE);

```
(If `action.sign` is guaranteed to be non-mutating when it fails, this could be after the call.)

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,548 @@
+                return None;
+            }
+
+            if !(merge_optional(&mut lhs.zkproof, zkproof)
+                && merge_optional(&mut lhs.spend_auth_sig, spend_auth_sig)
+                && merge_optional(&mut lhs.recipient, recipient)
+                && merge_optional(&mut lhs.value, value)
+                && merge_optional(&mut lhs.rcm, rcm)
+                && merge_optional(&mut lhs.rseed, rseed)
+                && merge_optional(&mut lhs.rcv, rcv)
+                && merge_optional(&mut lhs.proof_generation_key, proof_generation_key)
+                && merge_optional(&mut lhs.witness, witness)
+                && merge_optional(&mut lhs.alpha, alpha)
+                && merge_optional(&mut lhs.zip32_derivation, zip32_derivation)
+                && merge_optional(&mut lhs.dummy_ask, dummy_ask)
+                && merge_map(&mut lhs.proprietary, proprietary))
```
Here we can assume that if any of the fields on the LHS were `None`, then they were not needed to compute any sighash, and so there is no possibility of a merge invalidating a signature. It might be useful to point that out in a comment.

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,548 @@
+                    ),
+                };
+
+                if other_has_more_spends {
+                    self.spends.extend(spends.drain(self.spends.len()..));
+                }
+                if other_has_more_outputs {
+                    self.outputs.extend(outputs.drain(self.outputs.len()..));
+                }
```
This does not check whether we have any signatures (transparent or spend auth sigs). If we do then modifying either `self.spends` or `self.outputs` would invalidate them.

daira commented on pczt/src/sapling.rs:
```diff
@@ -0,0 +1,548 @@
+                return None;
+            }
+
+            if !(merge_optional(&mut lhs.zkproof, zkproof)
+                && merge_optional(&mut lhs.recipient, recipient)
+                && merge_optional(&mut lhs.value, value)
+                && merge_optional(&mut lhs.rseed, rseed)
+                && merge_optional(&mut lhs.rcv, rcv)
+                && merge_optional(&mut lhs.ock, ock)
+                && merge_optional(&mut lhs.zip32_derivation, zip32_derivation)
+                && merge_map(&mut lhs.proprietary, proprietary))
```
Here we can assume that if any of the fields on the LHS were `None`, then they were not needed to compute any sighash, and so there is no possibility of a merge invalidating a signature. It might be useful to point that out in a comment.

daira commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,535 @@
+    /// Entries are added by the Constructor, and modified by an Updater, IO Finalizer,
+    /// Signer, Combiner, or Spend Finalizer.
+    #[getset(get = "pub")]
+    pub(crate) actions: Vec<Action>,
```
This looks from `get = "pub"` as though it is only providing read access, but actually it generates an `actions_mut` accessor (and similarly for `{spends,outputs}_mut` on the Sapling bundle and  `{inputs,outputs}_mut` on the transparent bundle). I dislike the implicitness of `getset` tbh. I'm not saying we should get rid of it at this stage, but I'd have pushed back on it earlier if I'd realized it was implicitly generating mutable accessors. This makes it very difficult to determine whether invariants are preserved. In particular, the modifiability flags *don't* seem to be enforced. If there were explicit methods to modify these vectors, they could be.

str4d replied:
It doesn't implicitly generate mutable accessors; you have to explicitly request them with both a `derive(MutGetters)` and altering this to include `get_mut = "pub"`.

daira replied:
I think I was looking at a previous state; there was a force-push that removed `get_mut` iirc. But the main point is that there is mutation of the `{spends,outputs,actions}` vectors that is a) hard to audit, and b) doesn't check the flags, which violates [a "must" in BIP 370](https://github.com/bitcoin/bips/blob/master/bip-0370.mediawiki#constructor) if it's this crate's responsibility.

daira replied:
This thread was basically just an expression of my frustration at the incomplete `getset` documentation.

daira commented on pczt/src/transparent.rs:
```diff
@@ -0,0 +1,411 @@
+                return None;
+            }
+
+            if !(merge_optional(&mut lhs.sequence, sequence)
+                && merge_optional(&mut lhs.required_time_lock_time, required_time_lock_time)
+                && merge_optional(
+                    &mut lhs.required_height_lock_time,
+                    required_height_lock_time,
+                )
+                && merge_optional(&mut lhs.script_sig, script_sig)
+                && merge_optional(&mut lhs.redeem_script, redeem_script)
+                && merge_map(&mut lhs.partial_signatures, partial_signatures)
+                && merge_map(&mut lhs.bip32_derivation, bip32_derivation)
+                && merge_map(&mut lhs.ripemd160_preimages, ripemd160_preimages)
+                && merge_map(&mut lhs.sha256_preimages, sha256_preimages)
+                && merge_map(&mut lhs.hash160_preimages, hash160_preimages)
+                && merge_map(&mut lhs.hash256_preimages, hash256_preimages)
+                && merge_map(&mut lhs.proprietary, proprietary))
```
Here we can assume that if any of the fields on the LHS were `None`, then they were not needed to compute any sighash, and so there is no possibility of a merge invalidating a signature. It might be useful to point that out in a comment.

daira commented on pczt/src/transparent.rs:
```diff
@@ -0,0 +1,411 @@
+                return None;
+            }
+
+            if !(merge_optional(&mut lhs.redeem_script, redeem_script)
+                && merge_map(&mut lhs.bip32_derivation, bip32_derivation)
+                && merge_map(&mut lhs.proprietary, proprietary))
```
Here we can assume that if any of the fields on the LHS were `None`, then they were not needed to compute any sighash, and so there is no possibility of a merge invalidating a signature. It might be useful to point that out in a comment.

daira commented on pczt/src/roles/updater/transparent.rs:
```diff
@@ -0,0 +1,42 @@
+            .into_parsed()
+            .map_err(TransparentError::Parser)?;
+
+        bundle.update_with(f).map_err(TransparentError::Updater)?;
```
This doesn't check the modifiability flags.

str4d replied:
Those are irrelevant here, as the only permitted changes via an Updater are non-effecting.

daira replied:
I misread; this cannot cause mutability problems that are not addressed (or not, as the case may be!) by checks elsewhere.

daira commented on pczt/src/roles/updater/orchard.rs:
```diff
@@ -0,0 +1,38 @@
+
+        let mut bundle = orchard.into_parsed().map_err(OrchardError::Parser)?;
+
+        bundle.update_with(f).map_err(OrchardError::Updater)?;
```
This doesn't check the modifiability flags.

daira commented on pczt/src/roles/updater/sapling.rs:
```diff
@@ -0,0 +1,38 @@
+
+        let mut bundle = sapling.into_parsed().map_err(SaplingError::Parser)?;
+
+        bundle.update_with(f).map_err(SaplingError::Updater)?;
```
This doesn't check the modifiability flags.

str4d replied:
Those are irrelevant here, as the only permitted changes via an Updater are non-effecting.

daira reviewed (commented):
After having read BIP 370 thoroughly, I just don't understand the design of this API. [BIP 370](https://github.com/bitcoin/bips/blob/master/bip-0370.mediawiki#constructor) says, for the Constructor:
> Before any input or output may be added, the constructor must check the `PSBT_GLOBAL_TX_MODIFIABLE` field. Inputs may only be added if the Inputs Modifiable flag is True. Outputs may only be added if the Outputs Modifiable flag is True.

The adaptation of this rule to PCZTs is neither enforced by the API, nor is there enough information as it stands for a client of this API to reliably enforce it.

daira reviewed (dismissed):
utACK with comments.

str4d commented:
> After having read BIP 370 thoroughly, I just don't understand the design of this API. [BIP 370](https://github.com/bitcoin/bips/blob/master/bip-0370.mediawiki#constructor) says, for the Constructor:
> 
> > Before any input or output may be added, the constructor must check the `PSBT_GLOBAL_TX_MODIFIABLE` field. Inputs may only be added if the Inputs Modifiable flag is True. Outputs may only be added if the Outputs Modifiable flag is True.
> 
> That is not enforced by the API.

We don't provide an independent Constructor implementation, and thus have nowhere that needs to enforce this requirement. (If we did, we'd indeed need to enforce this requirement in that API.)

What we provide in our implemented API is `Builder::build_for_pczt`, which is inherently both Creator and Constructor. For the lifetime of `Builder`, the modifiability flags are implicitly `true`, and then once `Builder::build_for_pczt` is called, we have no API that enables adding subsequent Spends, Outputs, or Actions. (I know this because I did try implementing an independent Constructor, and quickly ran into the API safety rails we enforced in the `orchard` crate; bypassing these would have taken much longer and been more risky than instead baking PCZT support into the existing `Builder`s.)

nuttycom commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,548 @@
+        // If `bsk` is set on either bundle, the IO Finalizer has run, which means we
+        // cannot have differing numbers of actions, and the value sums must match.
+        match (self.bsk.as_mut(), bsk) {
+            (Some(lhs), Some(rhs)) if lhs != &rhs => return None,
+            (Some(_), _) | (_, Some(_))
```
What's going on with the scrutinee here? In the first and fourth cases, the type appears to be `(Option<_>, Option<_>)` in the second and third cases, it appears to be `((Option<_>, _), (Option<_>, _))`? How does this even compile?

nuttycom replied:
Oops, missed the `|`.

nuttycom commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,548 @@
+    /// Merges this bundle with another.
+    ///
+    /// Returns `None` if the bundles have conflicting data.
+    pub(crate) fn merge(
```
This should be a symmetric method `merge(left: Self, right: Self, ...)`; otherwise the asymmetry in `bsk` merging means you have to try the merge in both orders.

nuttycom replied:
nonblocking

nuttycom commented on pczt/src/orchard.rs:
```diff
@@ -0,0 +1,548 @@
+    /// Merges this bundle with another.
+    ///
+    /// Returns `None` if the bundles have conflicting data.
+    pub(crate) fn merge(
```
I note that at least according to the annotations, there doesn't seem to be test coverage of this method. If not adding tests that cover this in this PR, please open an issue.

daira replied:
We think this actually is covered. Perhaps codecov is not running the relevant integration test(s)?

nuttycom commented on pczt/src/roles/combiner/mod.rs:
The `combiner` role appears to be untested according to codecov; please include this in the "add missing PCZT tests" issue.

nuttycom commented on pczt/src/roles/creator/mod.rs:
```diff
@@ -0,0 +1,170 @@
+        let tx_version = match parts.version {
+            zcash_primitives::transaction::TxVersion::Sprout(_)
+            | zcash_primitives::transaction::TxVersion::Overwinter => None,
+            zcash_primitives::transaction::TxVersion::Sapling => Some(SAPLING_TX_VERSION),
```
I thought that only V5 transactions were supported by the underlying PCZT types? Why would the role support V4?

daira replied:
It's completely ok to generate a PCZT for a v4 tx, and we *might* be able to make that work later.

nuttycom commented on pczt/src/roles/prover/sapling.rs:
```diff
@@ -0,0 +1,45 @@
+use crate::Pczt;
+
+impl super::Prover {
+    pub fn create_sapling_proofs<S, O>(
```
Sapling proving appears to be untested.

daira commented on pczt/src/common.rs:
```diff
@@ -113,20 +150,26 @@ impl Global {
-        // - Bits 3-7 must be 0.
-        if (self.tx_modifiable >> 3) != 0 || (tx_modifiable >> 3) != 0 {
+        // - Bits 3-6 must be 0.
+        if ((self.tx_modifiable & !FLAG_SHIELDED_MODIFIABLE) >> 3) != 0
+            || ((tx_modifiable & !FLAG_SHIELDED_MODIFIABLE) >> 3) != 0
+        {
```
```suggestion
        if ((self.tx_modifiable & !(FLAG_TRANSPARENT_INPUTS_MODIFIABLE | FLAG_TRANSPARENT_OUTPUTS_MODIFIABLE | FLAG_SHIELDED_MODIFIABLE) != 0) {
```

nuttycom commented on zcash_primitives/CHANGELOG.md:
```diff
@@ -7,6 +7,17 @@ and this library adheres to Rust's notion of
 ## [Unreleased]
 
+### Added
+- `zcash_primitives::transaction`
```
I have not checked this for completeness; we should do a wholesale changelog review before release.

nuttycom reviewed (approved):
utACK 0cceac173ee2eabb935afa874e94a11967dce316; we should create an issue for adding some of the missing tests.

daira reviewed (approved):
utACK
